flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/writer/FileWriterBucket.java

```diff
@@ -0,0 +1,329 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.file.sink.writer;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.connector.file.sink.FileSink;
+import org.apache.flink.connector.file.sink.FileSinkCommittable;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.streaming.api.functions.sink.filesystem.BucketAssigner;
+import org.apache.flink.streaming.api.functions.sink.filesystem.BucketWriter;
+import org.apache.flink.streaming.api.functions.sink.filesystem.InProgressFileWriter;
+import org.apache.flink.streaming.api.functions.sink.filesystem.OutputFileConfig;
+import org.apache.flink.streaming.api.functions.sink.filesystem.RollingPolicy;
+
+import org.apache.commons.lang3.RandomStringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static org.apache.flink.streaming.api.functions.sink.filesystem.InProgressFileWriter.InProgressFileRecoverable;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * A bucket is the directory organization of the output of the {@link FileSink}.
+ *
+ * <p>For each incoming element in the {@code FileSink}, the user-specified
+ * {@link BucketAssigner} is queried to see in which bucket this element should be written to.
+ */
+@Internal
+class FileWriterBucket<IN, BucketID> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(FileWriterBucket.class);
+
+	private final BucketID bucketId;
+
+	private final Path bucketPath;
+
+	private final String uniqueId;
+
+	private final BucketWriter<IN, BucketID> bucketWriter;
+
+	private final RollingPolicy<IN, BucketID> rollingPolicy;
+
+	private final OutputFileConfig outputFileConfig;
+
+	private final List<InProgressFileWriter.PendingFileRecoverable> pendingFiles = new ArrayList<>();
+
+	private long partCounter;
+
+	@Nullable
+	private InProgressFileRecoverable inProgressFileToCleanup;
+
+	@Nullable
+	private InProgressFileWriter<IN, BucketID> inProgressPart;
+
+	/**
+	 * Constructor to create a new empty bucket.
+	 */
+	private FileWriterBucket(
+			BucketID bucketId,
+			Path bucketPath,
+			String uniqueId,
+			BucketWriter<IN, BucketID> bucketWriter,
+			RollingPolicy<IN, BucketID> rollingPolicy,
+			OutputFileConfig outputFileConfig) {
+		this.bucketId = checkNotNull(bucketId);
+		this.bucketPath = checkNotNull(bucketPath);
+		this.uniqueId = checkNotNull(uniqueId);
```

Comment:
Why not putting it here `RandomStringUtils.randomAlphanumeric(32)`? Do we need to set it deterministically somewhere for testing?

Reply:
Currently we did have no requirements for set it directly. I move the parameter to be initialized inside the `FileWriterBucket`.